src/SearchSystem/SearchBase.cs

```diff
@@ -0,0 +1,44 @@
+using System;
+using System.Collections.Generic;
+
+public abstract class SearchBase < M, E, T >
+{
+    protected M Repository;
+
+    public SearchBase(M manager)
+    {
+        this.Repository = manager;
+    }
+
+    public virtual void PerformSearch(T criterion)
```

Comment:
Don't you consider to make the use of bool type instead of void? And also when criterion is passed as null?